tmail-backend/blob/blobid-list/src/main/java/com/linagora/tmail/blob/blobid_list/memory/InMemorySingleSaveBlobStoreDAO.java

```diff
@@ -0,0 +1,41 @@
+package com.linagora.tmail.blob.blobid_list.memory;
+
+import com.google.common.collect.HashBasedTable;
+import com.google.common.collect.Table;
+import com.linagora.tmail.blob.blobid_list.BlobIdListManager;
+import org.apache.james.blob.api.BlobId;
+import org.apache.james.blob.api.BucketName;
+import org.apache.james.blob.api.ObjectStoreException;
+import org.apache.james.blob.memory.MemoryBlobStoreDAO;
+import reactor.core.publisher.Mono;
+
+public class InMemorySingleSaveBlobStoreDAO extends MemoryBlobStoreDAO {
+
+    private final Table<BucketName, BlobId, byte[]> blobs;
+    private final BlobIdListManager blobIdListManager;
+
+    public InMemorySingleSaveBlobStoreDAO(BlobIdListManager blobIdListManager) {
+        blobs = HashBasedTable.create();
+        this.blobIdListManager = blobIdListManager;
+    }
+
+    @Override
+    public Mono<Void> save(BucketName bucketName, BlobId blobId, byte[] data) {
+        return Mono.from(blobIdListManager.isStored(blobId))
+                .filter(isStore -> isStore.equals(false))
+                .doFirst(() -> {
+                    synchronized (blobs) {
```

Comment:
synchronized call in reactor should feel suspicious

Reply:
You can see this code at `src/main/java/org/apache/james/blob/memory/MemoryBlobStoreDAO.java:64`

Reply:
Crap. Let's say it is testing code then!